src/App.tsx

```diff
@@ -1,223 +1,264 @@
 /* eslint-disable jsx-a11y/control-has-associated-label */
-import React, { useContext, useEffect, useRef } from 'react';
+import React, {
+  useContext, useEffect, useMemo, useState,
+} from 'react';
 import { AuthContext } from './components/Auth/AuthContext';
+import {
+  getTodos, createTodo, updateMarkTodo, deleteTodo,
+} from './api/todos';
+import { Header } from './components/Header';
+import { TodoList } from './components/TodoList';
+import { ErrorMessage } from './components/ErrorMessage';
+import { Footer } from './components/Footer';
+import { Todo } from './types/Todo';
+import { Filter } from './types/Filter';
+import { ErrorType } from './types/ErrorType';
+import { Loading } from './types/Loading';
+
+let timerId = 0;
 
 export const App: React.FC = () => {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [tempTodo, setTempTodo] = useState(false);
+  const [todoTitle, setTodoTitle] = useState('');
+  const [error, setError] = useState<ErrorType>('');
+  const [filter, setFilter] = useState<Filter>('all');
+  const [isAdding, setIsAdding] = useState(false);
+  const [toggleAll, setToggleAll] = useState(false);
+  const [isLoading, setIsLoading] = useState<Loading>({});
+
   // eslint-disable-next-line @typescript-eslint/no-unused-vars
   const user = useContext(AuthContext);
-  const newTodoField = useRef<HTMLInputElement>(null);
 
-  useEffect(() => {
-    // focus the element with `ref={newTodoField}`
-    if (newTodoField.current) {
-      newTodoField.current.focus();
+  const getTodosFromServer = async () => {
+    if (!user) {
+      return;
     }
+
+    try {
+      const result = await getTodos(user.id);
+
+      setTodos(result);
+    } catch (err: unknown) {
+      throw new Error('Networ Error');
+    }
+  };
+
+  useEffect(() => {
+    getTodosFromServer();
   }, []);
 
+  useEffect(() => {
+    setToggleAll(todos.every(({ completed }) => completed === true));
+  }, [todos]);
+
+  const showError = (err: ErrorType) => {
+    clearTimeout(timerId);
+    setError(err);
+    timerId = window.setTimeout(() => setError(''), 3000);
+  };
+
+  const filteredByCompleted = (
+    isCompleted: boolean,
+    arrTodos = todos,
+  ): Todo[] => (
+    arrTodos.filter(todo => todo.completed === isCompleted)
+  );
+
+  const getIsLoading = (arrTodos: Todo[], loading: boolean) => {
+    const isLoadingObj: Loading = {};
+
+    arrTodos.forEach(({ id }) => {
+      isLoadingObj[id] = loading;
+    });
+
+    setIsLoading(isLoadingObj);
+  };
+
+  const itemsLeft = filteredByCompleted(false).length;
+
+  const completedTodos = filteredByCompleted(true);
```

Comment:
use `useMemo` for this variable, and `useCallback` for callbacks to avoid redundant recalculations